src/main/java/com/cloudogu/smeagol/wiki/infrastructure/GitClient.java

```diff
@@ -0,0 +1,150 @@
+        this.account = account;
+        this.repository = repository;
+        this.remoteUrl = remoteUrl;
+        this.branch = branch;
+    }
+
+    public void refresh() throws GitAPIException, IOException {
+        if ( repository.exists() ) {
+            pullChanges();
+        } else {
+            createClone();
+        }
+    }
+
+    private Git open() throws IOException {
+        if (git == null) {
+            git = Git.open(repository);
+        }
+        return git;
+    }
+
+    public File file(String path) {
+        return new File(repository, path);
+    }
+
+    public Optional<RevCommit> lastCommit(String path) throws IOException, GitAPIException {
+        Git git = open();
+        Iterator<RevCommit> iterator = git.log()
+                .addPath(path)
+                .setMaxCount(1)
+                .call()
+                .iterator();
+
+        if (iterator.hasNext()) {
+            return Optional.of(iterator.next());
+        }
+
+        return Optional.empty();
+    }
+
+    private void pullChanges()  throws GitAPIException, IOException {
+        LOG.trace("open repository {}", repository);
+        git = open();
+
+        LOG.debug("pull changes from remote for repository {}", repository);
+        git.pull()
+                .setRemote("origin")
+                .setRemoteBranchName(branch)
+                .setCredentialsProvider(credentialsProvider(account))
+                .call();
+    }
+
+    private void createClone()  throws GitAPIException, IOException {
+        LOG.info("clone repository {} to {}", remoteUrl, repository);
+        git = Git.cloneRepository()
+                .setURI(remoteUrl.toExternalForm())
+                .setDirectory(repository)
+                .setBranchesToClone(singleton("refs/head" + branch))
+                .setBranch(branch)
+                .setCredentialsProvider(credentialsProvider(account))
+                .call();
+
+        if (!"master".equals(branch)) {
+            File newRef = new File(repository, ".git/refs/heads/master");
+            File refDirectory = newRef.getParentFile();
+            if (!refDirectory.exists() && !refDirectory.mkdirs()) {
+                throw new IOException("failed to create parent directory " + refDirectory);
+            }
+            if (!newRef.exists() && !newRef.createNewFile()) {
+                throw new IOException("failed to create parent directory");
+            }
+            try (BufferedWriter output = new BufferedWriter(new FileWriter(newRef))) {
+                output.write("ref: refs/heads/" + branch);
+            }
+        }
+    }
+
+    private CredentialsProvider credentialsProvider(Account account) {
+        return new UsernamePasswordCredentialsProvider(account.getUsername(), account.getPassword());
+    }
+
+    public RevCommit commit(String path, String displayName, String email, String message) throws GitAPIException, IOException {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to throw at most one checked exception instead of: org.eclipse.jgit.api.errors.GitAPIException, java.io.IOException [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://ecosystem.cloudogu.com/sonar/coding_rules#rule_key=squid%3AS1160)